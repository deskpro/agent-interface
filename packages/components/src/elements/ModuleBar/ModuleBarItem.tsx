import * as React from "react";
import classNames from "classnames";

import Icon from "../Icon/Icon";

export type ModuleBarItemProps = {
  icon: string;
  label?: string;
  isActive?: boolean;
  onClick?: (e: React.MouseEvent<HTMLLinkElement>) => void;
  className?: string;
};

const ModuleBarItem: React.FC<ModuleBarItemProps> = ({
  icon,
  className,
  label,
  isActive = false,
  onClick,
  children
}) => (
  <li
    className={classNames("dp-ModuleBarList dp-ModuleBar-item", className, {
      "is-active": isActive
    })}
  >
    <Icon name={icon} size={21} onClick={onClick} />
    {!!label && <span className="dp-IconTooltip">{label}</span>}
    {children}
  </li>
);

export default ModuleBarItem;
