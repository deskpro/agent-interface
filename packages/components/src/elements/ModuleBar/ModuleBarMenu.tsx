import * as React from "react";
import classNames from "classnames";

import MenuItem from "./ModuleBarMenuItem";

type ModuleBarMenuSubcomponents = {
  Item: typeof MenuItem;
};

export type ModuleBarMenuProps = {
  className?: string;
  title: string;
  isVisible: boolean;
};

const ModuleBarMenu: React.FC<ModuleBarMenuProps> &
  ModuleBarMenuSubcomponents = ({ className, title, isVisible, children }) => (
  <div
    className={classNames("dp-ModuleBarMenu", className, {
      "is-visible": isVisible
    })}
  >
    <span className="dp-ModuleBarMenu-title">{title}</span>
    <ul>{children}</ul>
  </div>
);

ModuleBarMenu.Item = MenuItem;

export default ModuleBarMenu;
