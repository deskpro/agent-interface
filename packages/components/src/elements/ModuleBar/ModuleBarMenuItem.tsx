import * as React from "react";
import classNames from "classnames";

export type ModuleBarMenuItemProps = {
  className?: string;
  icon: "plane" | "plane-violet" | "plane-orange";
  title: string;
  subtitle?: string;
  link: string;
  linkComponent?: any;
  linkAttributes?: Partial<React.AnchorHTMLAttributes<HTMLAnchorElement>>;
  isActive?: any;
};

const ModuleBarMenuItem: React.FC<ModuleBarMenuItemProps> = ({
  className,
  icon,
  title,
  subtitle,
  isActive = false,
  link,
  linkComponent = "a",
  linkAttributes = {}
}) => (
  <li
    className={classNames("dp-ModuleBarMenu-item", className, {
      "is-active": isActive
    })}
  >
    {React.createElement(
      linkComponent,
      {
        ...linkAttributes,
        [typeof linkComponent === "string" && linkComponent === "a"
          ? "href"
          : "path"]: link
      },
      <span className={`dp-Icon Icon--${icon}`} />,
      <span>
        <span className="dp-ModuleBarMenu-itemTitle">{title}</span>,
        <span className="dp-ModuleBarMenu-itemSubtitle">{subtitle}</span>
      </span>
    )}
  </li>
);

export default ModuleBarMenuItem;
