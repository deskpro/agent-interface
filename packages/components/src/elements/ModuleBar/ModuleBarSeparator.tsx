import * as React from "react";

const ModuleBarSeparator: React.FC = () => (
  <li className="dp-ModuleBar-separator" />
);

export default ModuleBarSeparator;
