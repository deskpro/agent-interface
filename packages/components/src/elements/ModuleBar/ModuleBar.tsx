import * as React from "react";
import classNames from "classnames";
import ModuleBarItem from "./ModuleBarItem";
import ModuleBarSeparator from "./ModuleBarSeparator";

import "@deskpro/agent-interface-style/dist/components/dp-ModuleBar.css";

type ModuleBarSubcomponents = {
  Item: typeof ModuleBarItem;
  Separator: typeof ModuleBarSeparator;
};

export type ModuleBarProps = {
  className?: string;
};

const ModuleBar: React.FC<ModuleBarProps> & ModuleBarSubcomponents = ({
  className,
  children
}) => (
  <div className={classNames("dp-ModulebarContainer", className)}>
    <ul className="dp-ModuleBar">{children}</ul>
  </div>
);

ModuleBar.Item = ModuleBarItem;
ModuleBar.Separator = ModuleBarSeparator;

export default ModuleBar;
