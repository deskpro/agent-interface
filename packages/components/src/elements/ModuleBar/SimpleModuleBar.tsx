import * as React from "react";

import ModuleBar from "./ModuleBar";
import ModuleBarMenu from "./ModuleBarMenu";

export interface SimpleModuleBarItem {
  icon: string;
  label: string;
  name: React.Key;
}

export interface SimpleModuleBarPortalLink {
  icon: "plane" | "plane-violet" | "plane-orange";
  title: string;
  subtitle: string;
  link: string;
}

export type SimpleModuleBarProps = {
  className?: string;
  items: Array<SimpleModuleBarItem | "-">;
  portalMenuTitle?: string;
  portalMenuLinks?: SimpleModuleBarPortalLink[];
  onItemClick: (itemName: React.Key, e: React.SyntheticEvent) => void;
};

const SimpleModuleBar = ({
  className,
  items,
  onItemClick,
  portalMenuTitle = "Portals",
  portalMenuLinks = []
}) => {
  const [activeItem, setActiveItem] = React.useState<React.Key>();
  const [isPortalsMenuVisible, setMenuPortalsVisibility] = React.useState<
    boolean
  >(false);

  return (
    <ModuleBar className={className}>
      {items.map((item: SimpleModuleBarItem | "-", index: number) => {
        if (item === "-") {
          // eslint-disable-next-line react/no-array-index-key
          return <ModuleBar.Separator key={`separator_${index}`} />;
        }
        const { name, ...itemProps } = item;
        return (
          <ModuleBar.Item
            {...itemProps}
            key={name}
            isActive={activeItem === name}
            onClick={e => {
              setActiveItem(name);
              onItemClick(name, e);
            }}
          />
        );
      })}
      {!!portalMenuLinks.length && (
        <ModuleBar.Separator key="portal-separator" />
      )}
      {!!portalMenuLinks.length && (
        <ModuleBar.Item
          icon="portal"
          key="portal"
          onClick={() => setMenuPortalsVisibility(!isPortalsMenuVisible)}
        >
          <ModuleBarMenu
            title={portalMenuTitle}
            isVisible={isPortalsMenuVisible}
          >
            {portalMenuLinks.map((link: SimpleModuleBarPortalLink) => (
              <ModuleBarMenu.Item {...link} key={link.title} />
            ))}
          </ModuleBarMenu>
        </ModuleBar.Item>
      )}
    </ModuleBar>
  );
};

export default SimpleModuleBar;
