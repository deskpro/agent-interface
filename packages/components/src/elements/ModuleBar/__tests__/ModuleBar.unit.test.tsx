import * as React from "react";
import { render } from "react-testing-library";

import ModuleBar from "../ModuleBar";
import ModuleBarMenu from "../ModuleBarMenu";

describe("<TagBar />", () => {
  it("should match its snapshot", () => {
    const handleClick = jest.fn();
    const { container } = render(
      <ModuleBar className="module-bar">
        <ModuleBar.Item
          icon="ticket"
          label="Tickets"
          isActive
          onClick={handleClick}
        />
        <ModuleBar.Item
          icon="billing"
          label="Billing"
          isActive={false}
          onClick={handleClick}
        />
        <ModuleBar.Separator />
        <ModuleBar.Item icon="portal" onClick={handleClick}>
          <ModuleBarMenu title="Portals" isVisible>
            <ModuleBarMenu.Item
              icon="plane"
              title="Raven Paper"
              subtitle="support.ravenpaper.com"
              link="http://support.ravenpaper.com"
            />
            <ModuleBarMenu.Item
              icon="plane-violet"
              title="RP Europe"
              subtitle="support.rpeurope.com"
              link="http://support.rpeurope.com"
            />
          </ModuleBarMenu>
        </ModuleBar.Item>
      </ModuleBar>
    );
    expect(container.querySelector(".module-bar")).toMatchSnapshot();
  });
});
