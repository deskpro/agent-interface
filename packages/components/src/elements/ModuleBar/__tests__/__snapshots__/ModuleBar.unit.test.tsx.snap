// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TagBar /> should match its snapshot 1`] = `
<div
  class="dp-ModulebarContainer module-bar"
>
  <ul
    class="dp-ModuleBar"
  >
    <li
      class="dp-ModuleBarList dp-ModuleBar-item is-active"
    >
      <span
        class="dp-IconTooltip"
      >
        Tickets
      </span>
    </li>
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-IconTooltip"
      >
        Billing
      </span>
    </li>
    <li
      class="dp-ModuleBar-separator"
    />
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <div
        class="dp-ModuleBarMenu is-visible"
      >
        <span
          class="dp-ModuleBarMenu-title"
        >
          Portals
        </span>
        <ul>
          <li
            class="dp-ModuleBarMenu-item"
          >
            <a
              href="http://support.ravenpaper.com"
            >
              <span
                class="dp-Icon Icon--plane"
              />
              <span>
                <span
                  class="dp-ModuleBarMenu-itemTitle"
                >
                  Raven Paper
                </span>
                ,
                <span
                  class="dp-ModuleBarMenu-itemSubtitle"
                >
                  support.ravenpaper.com
                </span>
              </span>
            </a>
          </li>
          <li
            class="dp-ModuleBarMenu-item"
          >
            <a
              href="http://support.rpeurope.com"
            >
              <span
                class="dp-Icon Icon--plane-violet"
              />
              <span>
                <span
                  class="dp-ModuleBarMenu-itemTitle"
                >
                  RP Europe
                </span>
                ,
                <span
                  class="dp-ModuleBarMenu-itemSubtitle"
                >
                  support.rpeurope.com
                </span>
              </span>
            </a>
          </li>
        </ul>
      </div>
    </li>
  </ul>
</div>
`;
